core/store/hashtree/mem/tree.go

```diff
@@ -0,0 +1,477 @@
+		n.left = n.left.Delete(key)
+	} else {
+		n.right = n.right.Delete(key)
+	}
+
+	if n.left.GetType() == emptyNodeType && n.right.GetType() == emptyNodeType {
+		// If an interior node points to two empty nodes, it is itself an empty
+		// one.
+		return NewEmptyNode(n.depth)
+	}
+
+	return n
+}
+
+// Prepare implements mem.TreeNode. It updates the hash of the node and return
+// the digest.
+func (n *InteriorNode) Prepare(nonce []byte, prefix *big.Int, fac crypto.HashFactory) ([]byte, error) {
+	h := fac.New()
+
+	left, err := n.left.Prepare(nonce, new(big.Int).SetBit(prefix, int(n.depth), 0), fac)
+	if err != nil {
+		// No wrapping to prevent recursive calls to create huge error messages.
+		return nil, err
+	}
+
+	right, err := n.right.Prepare(nonce, new(big.Int).SetBit(prefix, int(n.depth), 1), fac)
+	if err != nil {
+		// No wrapping to prevent recursive calls to create huge error messages.
+		return nil, err
+	}
+
+	_, err = h.Write(append(left, right...))
+	if err != nil {
+		return nil, xerrors.Errorf("interior node failed: %v", err)
+	}
+
+	n.hash = h.Sum(nil)
+
+	return n.GetHash(), nil
+}
+
+// Visit implements mem.TreeNode. It executes the callback with the node and
+// recursively with the children.
+func (n *InteriorNode) Visit(fn func(TreeNode)) {
+	fn(n)
+	n.left.Visit(fn)
+	n.right.Visit(fn)
+}
+
+// Clone implements mem.TreeNode. It returns a deep copy of the interior node.
+func (n *InteriorNode) Clone() TreeNode {
+	return &InteriorNode{
+		depth: n.depth,
+		left:  n.left.Clone(),
+		right: n.right.Clone(),
+	}
+}
+
+// LeafNode is a leaf node with a key and a value.
+type LeafNode struct {
+	hash  []byte
+	depth uint16
+	key   []byte
+	value []byte
+}
+
+// NewLeafNode creates a new leaf node.
+func NewLeafNode(depth uint16, key, value []byte) *LeafNode {
+	return &LeafNode{
+		depth: depth,
+		key:   key,
+		value: value,
+	}
+}
+
+// GetHash implements mem.TreeNode. It returns the hash of the node.
+func (n *LeafNode) GetHash() []byte {
+	return append([]byte{}, n.hash...)
+}
+
+// GetType implements mem.TreeNode. It returns the leaf node type.
+func (n *LeafNode) GetType() byte {
+	return leafNodeType
+}
+
+// Search implements mem.TreeNode. It returns the value if the key matches.
+func (n *LeafNode) Search(key *big.Int, path *Path) []byte {
+	if path != nil {
+		path.leaf = n
+	}
+
+	if makeKey(n.key).Cmp(key) == 0 {
+		return n.value
+	}
+
+	return nil
+}
+
+// Insert implements mem.TreeNode. It replaces the leaf node by an interior node
+// that contains both the current pair and the new one to insert.
+func (n *LeafNode) Insert(key, value []byte, bi *big.Int) TreeNode {
+	curr := makeKey(n.key)
+
+	if curr.Cmp(bi) == 0 {
+		n.value = value
+		return n
+	}
+
+	node := NewInteriorNode(n.depth)
+
+	if curr.Bit(int(n.depth)) == 0 {
+		node.left = node.left.Insert(n.key, n.value, curr)
+	} else {
+		node.right = node.right.Insert(n.key, n.value, curr)
+	}
+
+	if bi.Bit(int(n.depth)) == 0 {
```

Comment:
what if both `if` are true? Is it always mutually exclusive ?

Reply:
Yes it is

Reply:
ok, so maybe we could merge those two ifs? Or maybe simply a comment?

Reply:
You definitely can't merge them, it would be wrong. But a comment is feasible!